VoodooPS2Trackpad/VoodooPS2Elan.h

```diff
@@ -0,0 +1,365 @@
+/*
+ * Elan PS2 touchpad integration
+ *
+ * Mostly contains code ported from Linux
+ * https://github.com/torvalds/linux/blob/master/drivers/input/mouse/elantech.c
+ *
+ * Created by Bartosz Korczyński (@bandysc), Hiep Bao Le (@hieplpvip)
+ * Special thanks to Kishor Prins (@kprinssu), EMlyDinEsHMG and whole VoodooInput team
+ */
+
+#ifndef _APPLEPS2ELAN_H
+#define _APPLEPS2ELAN_H
+
+#include "../VoodooPS2Controller/ApplePS2MouseDevice.h"
+#include "LegacyIOHIPointing.h"
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Winconsistent-missing-override"
+#include <IOKit/IOCommandGate.h>
+#include <IOKit/acpi/IOACPIPlatformDevice.h>
+#pragma clang diagnostic pop
+
+#include "VoodooInputMultitouch/VoodooInputEvent.h"
+
+struct virtual_finger_state {
+    TouchCoordinates prev;
+    TouchCoordinates now;
+    uint8_t pressure;
+    uint8_t width;
+    bool touch;
+    bool button;
+    MT2FingerType fingerType;
+};
```

Comment:
Are you sure that averaging is not needed? Does the cursor shake if a finger is placed on the touchpad and doesn't move?

Reply:
I have hardware v4 and everything works fine for me. Hardware v4 reports coordinates for every finger (up to 5), unlike Synaptics which reports only two. I believe averaging/number renumbering is not needed for v4 but needed for older hardware. I don't have older hardware nor I understand your algorithm so I don't implement it.

The limitation for older hardware is:
- v1: Only the number of fingers (max 3) and coordinates of one of them is reported
- v2: Maximum number of fingers is 4. For one/two fingers touch, all coordinates are reported. For three/four fingers touch, only coordinates of the "center finger" is reported.
- v3: Same as v2, except the maximum number of fingers is 3.

Currently, v1 and three+ fingers touch for v2/v3 is handled by generating fake fingers. Not sure if it's a good approach.

Reply:
By averaging I mean not calculating coordinates for fake fingers, but smoothing for each of reported fingers based on previous coordinates (that's what `SimpleAverage` and `DecayingAverage` classes do). On Synaptics, without smoothing the cursor would shake if a finger is not moving. It's not only annoying, but also makes 'tap to click' gesture very difficult because macOS thinks that the cursor is moving.

Reply:
I don't have that problem

Reply:
Neither have I, for me (elan v4) touch is very stable even without any averaging